Symmetry and Data Augmentation

Scienfitz opened the pull request:
Implements #621 

- New `Symmetry` class which owns augmentation logic and acts as the activator for related mechanisms by presence
- Adding a symmetry to a Bayesian recommender will perform surrogate-agnostic data augmentation
- Future plans: Adding a symmetry to a surrogate will trigger architectural changes that incorporate the symmetries (eg kernels in a GP). There is just a dummy implementation of this in this PR so far
- New docs
  - [Example](https://scienfitz.github.io/baybe-dev/latest/examples/Symmetries/permutation.html) demonstrating the powerful impact of data augmentation in the presence of a invariances in the searchspace
  - [Components](https://scienfitz.github.io/baybe-dev/latest/components/symmetries.html) page about the symmetries
  - [Mentioning](https://scienfitz.github.io/baybe-dev/latest/components/recommenders.html#data-augmentation) of the recommender-based augmentation activation

**Notes**
- There are some inconsistencies with the way constraints and symmetries are designed. These will be rectified in the constraint refactoring #847 
- Unrelated bugfix: When creating the example it was realized that the automatic application of a no-label-duplicates constraint within the permutation constraint is wrong. This has thus been removed. It should not pose a major problem because for mixture use cases the application of both invariant and label-dedupe constraints was performed (meaning the label deduplication was applied twice, yielding the correct result anyway)

Scienfitz commented on docs/scripts/build_examples.py:
```diff
@@ -44,7 +48,7 @@ def build_examples(destination_directory: Path, dummy: bool, remove_dir: bool):
     # This list contains the order of the examples as we want to have them in the end.
     # The examples that should be the first ones are already included here and skipped
-    # later on. ALl other are just included.
+    # later on. All others are just included.
```
thx lol

Scienfitz commented:
FYI I'm putting this back to draft as it makes sense to incorporate the symmetry concept already now. Things like the example will only minimally change though so that can already be looked at.

Scienfitz commented on baybe/symmetries.py:
```diff
@@ -0,0 +1,439 @@
+
+    # Object variables
+    # TODO: Needs inner converter to tuple
+    copermuted_groups: tuple[tuple[str, ...], ...] = field(
```
This is the only exception where I did still somewhat follow the design of the corresponding constraint. Because all cosntraints share `parameters` this symmetry requires to specify `copermuted_groups` as separate attribute. But actually all the groups could be combined into a single attribute.

Instead of `parameter_names=["a1", "a2"]` and `copermuted_groups=[["b1", "b2"],["c1","c2"]]` we would have `permutation_groups = [["a1","a2"], ["b1", "b2"], ["c1","c2"]]`. This is certainly more lgical but maybe a bit less userfirendly if there is really just 1 group.

This would have as consequence that we need to (re)define what the `parameters` property even should contain or whether it should even be dropped.

In the absence of opinions this wills tay as it is.

Scienfitz replied:
its not the same because the copermuted groups are not permuted independent of the primary group, but in exactly the same way as the primary group. Imagine it like 
- taking an iterable (the primary group)
- then get all permutations of the index of that iterable
- create all permutations of the original iterable by applying all permuted indices
- Now you also apply the permuted indices to other iterables (called copermuted groups)

As I wrote in my comment I would be open to get rid of the primary and copermuted groups by combining them into one single attribute, but it would have a very different interface for Symemtries as consequence (where we even have to debate whether Symmetries (and bye xtension also constraints)) can have a propert like `.parameter_names` and what it would mean

Scienfitz commented on tests/hypothesis_strategies/symmetries.py:
```diff
@@ -0,0 +1,123 @@
+    n_copermuted_groups = draw(
+        st.integers(
+            min_value=0,
+            max_value=len(parameter_names_pool) // len(parameter_names_pool),
```
@copilot The line is completely botched, should probably be `max_value = len(parameter_names_pool) // n_params - 1`

AdrianSosic commented on baybe/constraints/discrete.py:
```diff
@@ -338,6 +350,18 @@ def get_invalid(self, data: pd.DataFrame) -> pd.Index:
+
+        return PermutationSymmetry(
+            parameter_names=self.parameters,
+            copermuted_groups=(tuple(self.dependencies.parameters),)  # type: ignore[arg-type]
```
* why is there an explicit tuple conversion here 
* why do we need the type silencing

--> Neither should be required when the field converters are properly set up

Scienfitz replied:
I mean the attribute expects a tuple, so I am providing a tuple. Are you criticising this?

The type silence here was something that I actively tried to fix but that seems to be very hard. Now I dont remember the details tho I would have to check again

AdrianSosic commented on baybe/utils/conversion.py:
```diff
@@ -42,6 +42,18 @@ def nonstring_to_tuple(x: Sequence[_T], self: type, field: Attribute) -> tuple[_
     return tuple(x)
 
 
+def normalize_str_sequence(
```
The newly introduced name is misleading and the introduced type hint is false, because:
you introduced this function when we enabled Booleans for categorical parameters. Both name and type hint suggest that the incoming object is already a sequence of strings, which is not the case – it's more generally a sequence of "convertible to string", and right now we use it for str/bool.

AdrianSosic commented on baybe/symmetries.py:
```diff
@@ -0,0 +1,439 @@
+    use_data_augmentation: bool = field(
+        default=True, validator=instance_of(bool), kw_only=True
+    )
+    """Flag indicating whether data augmentation would be used with surrogates that
```
Two questions/remarks:
* `would` sounds a bit weird, no? How about `is to be`
* Why the `that support this`? Isn't the beauty of the augmentation approach that it's surrogate-agnostic?

AdrianSosic commented on baybe/symmetries.py:
```diff
@@ -0,0 +1,439 @@
+
+    @abstractmethod
+    def augment_measurements(
+        self, df: pd.DataFrame, parameters: Iterable[Parameter]
```
previously there was `data`, now it's `df` --> let's try to use `measurements` wherever we talk about a "measurements dataframe". I think we should also add a doc page at some point (not this PR) where the format is explained. Currently, this is done in one of the core entry methods, but I think the doc would be a nicer place and we then simply link from all methods to that place.

AdrianSosic commented on baybe/symmetries.py:
```diff
@@ -0,0 +1,439 @@
+    )
+    """The names of the parameters affected by the symmetry."""
+
+    @override
```
The definition order is a bit messed up here, because the property is in between attributes.

Let's stick with the following pattern (where initialization-related stuff comes first)
* attributes
* validators / post-init
* properties
* methods

AdrianSosic commented on baybe/symmetries.py:
```diff
@@ -0,0 +1,439 @@
+    # Object variables
+    # TODO: Needs inner converter to tuple
+    copermuted_groups: tuple[tuple[str, ...], ...] = field(
+        factory=tuple, converter=tuple
```
I guess this needs itself a `cattrs` converter, since we otherwise force people to pass things as tuple

AdrianSosic commented on baybe/symmetries.py:
```diff
@@ -0,0 +1,439 @@
+            searchspace.parameter_names
+        )
+        if parameters_missing:
+            raise ValueError(
```
Is `IncompatibilityError` maybe more appropriate to raise in these validation methods (also in subclasses)? Because it indicates the the symmetry is `incompatible` with the search space, whereas `ValueError` would be more for things `within` the same class?

AdrianSosic commented on baybe/symmetries.py:
```diff
@@ -0,0 +1,439 @@
+    A dependency symmetry expresses that certain parameters are dependent on another
+    parameter having a specific value. For instance, the situation "The value of
+    parameter y only matters if parameter x has the value 'on'.". In this scenario x
+    is the causing parameter and y depends on x.
```
Comment regarding terminology, here and in other docstrings: since the name is  "DependencySymmetry" and we talk about the "dependent parameters":
* Wouldn't make sense to call the "causing" parameter the "independent parameter" instead? Would be more consistent and also solves the problem that "causing" sort of implies a causal relationship from x to y (in the sense that x physically affects y) which is not the case
* You call them "dependent" parameters, which is fine to me, but why do you then call the corresponding attribute "affected"?

Scienfitz replied:
@AdrianSosic in the absence of any reply and me not being particularly keen on doing more naming changes as part of this PR I am closing

AdrianSosic reviewed (changes requested):
Hi @Scienfitz, here my first batch of comments. Still need to finalize the dependency symmetry and the docs, but my github UI is getting so slow that I feel I need to submit the pending comments 😬

AVHopp commented on examples/Symmetries/permutation.py:
```diff
@@ -0,0 +1,216 @@
+# # Optimizing a Function that is Permutation Invariant
```
```suggestion
# # Optimizing a Permutation-Invariant Function
```

AVHopp commented on examples/Symmetries/permutation.py:
```diff
@@ -0,0 +1,216 @@
+
+# There are several ways to handle such symmetries. The simplest one is
+# to augment your data. In the case of permutation invariance, augmentation means for
+# each measurement (x,y) you also add a measurement with switched values: (y,x).
```
```suggestion
# each measurement $(x,y)$ you also add a measurement with switched values, i.e.,  $(y,x)§.
```

AVHopp commented on examples/Symmetries/permutation.py:
```diff
@@ -0,0 +1,216 @@
+axs[0].set_ylabel("y");
+# fmt: on
+
+# The first subplot shows the function we want to minimize. The dashed red line
```
You should mention that this is to be found at the end of the file.

AVHopp commented on examples/Symmetries/permutation.py:
```diff
@@ -0,0 +1,216 @@
+# fmt: on
+
+# The first subplot shows the function we want to minimize. The dashed red line
+# illustrates the permutation invariance, which is similar to a mirror-symmetry, just
```
Is the function actually perfectly permutation invariant? At least in the image, some things seem to be a little bit off. Is this just an artifact of the plotting?

<img width="45" height="42" alt="Image" src="https://github.com/user-attachments/assets/4ff56384-f4f2-42eb-8d74-862112c56335" />

AVHopp commented on examples/Symmetries/permutation.py:
```diff
@@ -0,0 +1,216 @@
+# including augmented ones. So as a user, you don't have to generate permutations and
+# add them manually. Depending on the surrogate model, this might have different
+# impacts. We can expect a strong effect for tree-based models because their splits are
+# always parallel to the parameter axes. Thus, without augmented measurements, it is
```
```suggestion
# impacts. For example, we can expect a strong effect for tree-based models because their splits are
# always parallel to the parameter axes. Thus, without augmented measurements, it is
```

AVHopp replied:
Reason: Otherwise, this reads as if a full list of the impact on different models might come next.

AVHopp commented on examples/Symmetries/permutation.py:
```diff
@@ -0,0 +1,216 @@
+objective = NumericalTarget("f", minimize=True).to_objective()
+
+# We set up a constrained and unconstrained searchspace to demonstrate the impact of
+# the constraint on optimization performance.
```
```suggestion
# We set up a constrained and an unconstrained searchspace to demonstrate the impact of
# the constraint on optimization performance.
```

AVHopp replied:
Maybe even mention the word "plain" here such that the description is closer the the code?

AVHopp commented on examples/Symmetries/permutation.py:
```diff
@@ -0,0 +1,216 @@
+print(f"{'Without Constraint:':<35} {len(searchspace_plain.discrete.exp_rep)}")
+print(f"{'With Constraint:':<35} {len(searchspace_constrained.discrete.exp_rep)}")
+
+# We can see that the searchspace without the constraint has more points than the other
```
Naively, I would have expected a factor of 2 as the difference. It is clear that this factor is not met explicitly due to the diagonal to me, but I'm asking myself if we should maybe add a comment like "We can see that the searchspace without the constraint has roughly doubler the number of points compared to the unconstrained one. This is exactly what we expect, as we only model one half of the `parameter triangle`. Note that it is not a factor of exactly 2 due to the points on the diagonal"

AVHopp commented on examples/Symmetries/permutation.py:
```diff
@@ -0,0 +1,216 @@
+# subplot.
+
+# BayBE can automatically perform this augmentation if configured to do so.
+# Specifically, surrogate models have the `Surrogate.symmetries` attribute. If any of
```
Any reason why `Surrogate.symmetries` is not a link?

AVHopp commented on examples/Symmetries/permutation.py:
```diff
@@ -0,0 +1,216 @@
+
+symmetry = constraint.to_symmetry(use_data_augmentation=True)
+recommender_plain = TwoPhaseMetaRecommender(
+    recommender=BotorchRecommender(surrogate_model=NGBoostSurrogate())
```
Should we add an explanation on why we do not use the default surrogate model?

Scienfitz replied:
afaik there is commentary on the choice of model in the example

AVHopp reviewed (commented):
First set of comments, only on user guide and example without having looked at the code. Like them both quite a lot, comments are mostly minor.

AVHopp reviewed (commented):
Incomplete, but cannot finish today

AVHopp commented on baybe/symmetries.py:
```diff
@@ -0,0 +1,439 @@
+
+                ref_vals = set(cast(DiscreteParameter, params[0]).values)
+                if any(
+                    set(cast(DiscreteParameter, p).values) != ref_vals for p in params
```
Couldn't you start at `params[1]` here as the first comparison will be between `params[0]` and `ref_vals` which is the same?

AVHopp commented on baybe/symmetries.py:
```diff
@@ -0,0 +1,439 @@
+    """The names of the causing parameter others are depending on."""
+
+    # object variables
+    condition: Condition = field(validator=instance_of(Condition))
```
Currently, the docstring of the `Condition` class is

"""Abstract base class for all conditions.

    Conditions always evaluate an expression regarding a single parameter.
    Conditions are part of constraints, a constraint can have multiple conditions.
"""

This should probably be adjusted since they are now also part of `Symmetry`

AVHopp commented on baybe/symmetries.py:
```diff
@@ -0,0 +1,439 @@
+    """The parameters affected by the dependency."""
+
+    n_discretization_points: int = field(
+        default=3, validator=(instance_of(int), ge(2)), kw_only=True
```
Any specific reason for this as the default choice?

Scienfitz replied:
we need a default because its not relevant for cases where there is no conti parmeter

given we need a default I am picking a samll number that wouldnt make everything explode in case there are several discretizations needed

AVHopp reviewed (commented):
Review is still WIP, but since I won't be able to add more comments soon here is the first batch

Scienfitz commented on examples/Symmetries/permutation.py:
```diff
@@ -0,0 +1,216 @@
+# There are several ways to handle such symmetries. The simplest one is
+# to augment your data. In the case of permutation invariance, augmentation means for
+# each measurement (x,y) you also add a measurement with switched values: (y,x).
+# This has the advantage that it is fully model-agnostic, but might
```
`it` refers to the same thing as `This`, ie the method literally discussed int he sentence immediately preceding this one

Scienfitz commented on baybe/symmetries.py:
```diff
@@ -0,0 +1,439 @@
+        Args:
+            df: The dataframe containing the measurements to be augmented.
+            parameters: Parameter objects carrying additional information (might not be
+                needed by all augmentation implementations).
```
right I should give it None as default

Scienfitz commented on baybe/symmetries.py:
```diff
@@ -0,0 +1,439 @@
+    $f(x,y) = f(y,x)$.
+    """
+
+    _parameter_names: tuple[str, ...] = field(
```
its very similar to `targets` of obejctives. some just have a single target so we let objectives store the amount in either a single entry or a collection depending on their choice. However, all sare the common `targets` property

same idea here

Scienfitz commented on baybe/symmetries.py:
```diff
@@ -0,0 +1,439 @@
+
+    # Object variables
+    # TODO: Needs inner converter to tuple
+    copermuted_groups: tuple[tuple[str, ...], ...] = field(
```
open for better names

copermuted = a group permuted in conjuncture (ie the same way) with the primary group NOT an independently permuted group

Scienfitz commented on baybe/symmetries.py:
```diff
@@ -0,0 +1,439 @@
+        #  - copermuted_groups = [["a_1", "a_2", ...], ["b_1", "b_2", ...]]
+        # indicating that the groups "a_k" and "b_k" should be permuted in the same way
+        # as the group "p_k".
+        # We create `groups` to look like (("p1", "a1", "b1"), ("p2", "a2", "b2"), ...).
```
this is creating entries such that they can be readily passed to the utility
I cannot speak whether thats the same as a mathematican would see it, but does it really matter?

AVHopp replied:
I'm not aware of the concept of a co-permutation in mathematics and the corresponding formalization. Probably, mathematicians would model such a situation in general differently. But let's first see what the others say here - if I am alone with my interpretation, I am happy to simply resolve. If not, we can dig deeper into this.

AVHopp replied:
I disagree with you claming that I have no point here. I was referring to this part of the description that you give:

`We create "groups" to look like (("p1", "a1", "b1"), ("p2", "a2", "b2"), ...).`

All that I say is that THIS notation could be interpreted as "I map p* to a* and a* to b* and b* to p*" as this exactly the way that permutations are noted in mathematics when using Cycle notation: https://en.wikipedia.org/wiki/Permutation#Cycle_notation and this is NOT what you mean. In cycle notation, you would probably write this down as `(p1, p2, p3), (a1, a2, a3), (b1, b2, b3)`. Maybe this is also the solution to my confusion about the permuted group and the co-permuted groups: It seems like those are simply different cycles of the same permutation.

Maybe a simple comment "Note that this is not the cycle notation of the permutation" or like this might be sufficient, but as I said, let's see what the others say here.

Scienfitz commented on baybe/utils/augmentation.py:
```diff
@@ -188,3 +178,66 @@ def df_apply_dependency_augmentation(
+            row_new[column] = 2.0 * mirror_point - row[column]
+            to_add.append(row_new)
+
+        # Store augmented rows, keeping the index of their original row
```
for applicaiton purposes its not really relevant as the index doesnt matter

but if you debug this its a useful info to have because you can immediately assign the augmented points to their "original" one

Given that its trivial to keep the index here I see no arg to remove it

Scienfitz commented on baybe/symmetries.py:
```diff
@@ -0,0 +1,439 @@
+    $f(x,y) = f(y,x)$.
+    """
+
+    _parameter_names: tuple[str, ...] = field(
```
do you mean the one on the `Converter` line?

I just took this pattern from existing parts of the code I assume that that pattern is a bit tricky for mypy but don't know the details

AVHopp replied:
I think that this was what I meant, but don't remember either. IN that case I'll resolve.

Scienfitz commented on baybe/symmetries.py:
```diff
@@ -0,0 +1,439 @@
+    """Abstract base class for symmetries.
+
+    Symmetry is a concept that can be used to configure the modelling process in the
+    presence of invariances.
```
`Symmetry` on its own is certainly a concept, just like a `Parameter` is a concept

AVHopp commented on baybe/symmetries.py:
```diff
@@ -0,0 +1,439 @@
+        # indicating that the groups "a_k" and "b_k" should be permuted in the same way
+        # as the group "p_k".
+        # We create `groups` to look like (("p1", "a1", "b1"), ("p2", "a2", "b2"), ...).
+        # It results in just (("p1",), ("p2",), ...) if there are no copermuted groups.
```
Ok, then please adjust the docstring and feel free to simply resolve here afterwards.

Scienfitz committed:
Refactor shared validators and converters

Scienfitz committed:
Update permutation augmentation utility interface

Scienfitz committed:
Add mirror augmentation utility

Scienfitz committed:
Add Symmetry domain model

Scienfitz committed:
Add Parameter.is_equivalent and apply in PermutationSymmetry

Scienfitz committed:
Integrate symmetries into surrogates and recommenders

Scienfitz committed:
Update constraints for symmetry support

Scienfitz committed:
Add hypothesis strategies for symmetries and conditions

Scienfitz committed:
Add symmetry tests

Scienfitz committed:
Add symmetry documentation

Scienfitz committed:
Add symmetry example

Scienfitz committed:
Handle CompositeSurrogate in symmetry integration

The _autoreplicate converter on main wraps surrogates in a
CompositeSurrogate. Access the inner template for symmetry
validation and augmentation.

Scienfitz committed:
Fix mypy errors in categorical validator and dependency type ignore

Scienfitz committed:
Add symmetry validation tests

Scienfitz committed:
Update CHANGELOG

Scienfitz committed:
Replace deprecated set_random_seed with Settings in example

Scienfitz committed:
Fix Sphinx cross-references for symmetry classes

Use full module paths (e.g., baybe.symmetries.base.Symmetry)
instead of short paths via __init__.py re-exports, which
Sphinx cannot resolve.

Scienfitz commented:
@AdrianSosic @AVHopp this PR is again ready for review

I
- inocorporated most content arising from your comments
- then made a new branch off of `main`
- replayed all end2end changes the old branch had into a fresh commit structure on the new branch
- overwrote the old branch with the new one

Scienfitz committed:
Fix bug in permutation constraint

`DiscretePermutationInvarianceConstraint` was always internally applying a DiscreteNoLabelDuplicates constraint to remove the diagonal elements, which is not correct and can always be achieved separately by explicitly using `DiscreteNoLabelDuplicates`

AVHopp reviewed (commented):
First batch of comments, not yet done

Scienfitz committed:
Improve docstring

Co-authored-by: Alexander V. Hopp <alexander.hopp@merckgroup.com>

Scienfitz committed:
Add docstrings to to_symmetries and to_symmetry methods

Scienfitz committed:
Add use_data_augmentation to symmetry summary

Scienfitz committed:
Rework imports

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/emdgroup/baybe?pullRequest=626) <br/>All committers have signed the CLA.

Scienfitz commented on baybe/surrogates/gaussian_process/core.py:
```diff
@@ -142,7 +142,7 @@ def _posterior(self, candidates_comp_scaled: Tensor, /) -> Posterior:
     @override
     def _fit(self, train_x: Tensor, train_y: Tensor) -> None:
-        import botorch
+        import botorch.models.transforms
```
i think it was working becuase of the way the top-level `__init__` spaces are invoked even if it seems this is just importoing `transforms` 

but I agree that its confusing and looks a bit odd, I reworked the imports here to be very exlicit be1c293b552b02b0385f0d7cac6598787d0ec329

Scienfitz commented on baybe/symmetries/base.py:
```diff
@@ -0,0 +1,87 @@
+    def parameter_names(self) -> tuple[str, ...]:
+        """The names of the parameters affected by the symmetry."""
+
+    def summary(self) -> dict:
```
overlooked indeed and fixed in 9f076b6762ea7026b17a4aa6d303c5295ff5d09d

Scienfitz commented on baybe/constraints/discrete.py:
```diff
@@ -272,39 +271,42 @@ def get_invalid(self, data: pd.DataFrame) -> pd.Index:
+    def to_symmetries(
+        self, use_data_augmentation=True
+    ) -> tuple[DependencySymmetry, ...]:
+        """Convert to a :class:`~baybe.symmetries.dependency.DependencySymmetry`."""
```
added docstrings c853cac4cd2f994a292f236691489145df626029

Scienfitz commented on baybe/recommenders/pure/bayesian/base.py:
```diff
@@ -114,6 +130,13 @@ def _setup_botorch_acqf(
 
+        # Perform data augmentation if configured
+        surrogate_for_augmentation = self._get_surrogate_for_augmentation()
+        if surrogate_for_augmentation is not None:
```
I cant fully follow you here, can you elaborate?

this is not just a check, it combines retrieving the surrogate to call the augmentation
the check you desribe is actually 100% redundant as `augment_measurements` will just run over an empty list (so we cann just call it independent of whether there are symmetries configured or not)

AVHopp replied:
I also don't remember anymore what I wanted here :D

AVHopp commented on docs/userguide/symmetries.md:
```diff
@@ -0,0 +1,51 @@
+| Symmetry                                       | Functional Definition                                                                                                                            | Corresponding Constraint                                                                           | 
+|:-----------------------------------------------|:-------------------------------------------------------------------------------------------------------------------------------------------------|:---------------------------------------------------------------------------------------------------|
+| {class}`~baybe.symmetries.permutation.PermutationSymmetry` | $f(x,y) = f(y,x)$                                                                                                                                | {class}`~baybe.constraints.discrete.DiscretePermutationInvarianceConstraint`                       |
+| {class}`~baybe.symmetries.dependency.DependencySymmetry`  | $f(x,y) = \begin{cases}f(x,y) & \text{if }c(x) \\f(x) & \text{otherwise}\end{cases}$<br>where $c(x)$ is a condition that is either true or false | {class}`~baybe.constraints.discrete.DiscreteDependenciesConstraint`                                |
```
I do believe that functions can work like that, I just found it a bit hard to parse. If I find a better formulation I will post one here. If you don't get a reply here until the end of the week, feel free to simply resolve.

AVHopp replied:
Can we maybe add some horizontal lines here between the different types of symmetries? I find it a bit hard too read (but that might be personal preference/opinion) 

<img width="944" height="273" alt="Image" src="https://github.com/user-attachments/assets/ef8b4c42-31d0-477c-83b6-072c932fc0d3" />

AVHopp replied:
Formally, one could write it down like that: 
```Given a function $f:X\times Y \to \mathbb{R}$ as well as a condition function $c:X\to\{True, False\}$, we say that $f$ has a dependency symmetry on $f$ if and only if for all $x\in X$, it holds that $\neg c(x) \implies \forall y_1, y_2 \in Y : f(x, y_1) = f(x, y_2)=: f(x)$ ```

However, that is quite verbose, and I can see that this maybe too much. Let me summon @AdrianSosic here, I would also be fine with using the current variant.

AVHopp replied:
Rendered variant for your convenience: 

<img width="614" height="116" alt="Image" src="https://github.com/user-attachments/assets/e1baf3d9-2480-4042-8ce6-bcf93ad91895" />

Scienfitz replied:
really not a fan of this

we are not writing math papers we are trying to convey ideas, there is no requirement for such overblown formality here

Here is an example, that is extremley similar to what I wrote
<img width="325" height="74" alt="image" src="https://github.com/user-attachments/assets/1f50faec-98d5-4fb5-a183-74c1147680ba" />
it follows the same idea just here we are reducing by two variables instead of one if the condition is not met -  would you have complained about this as well?

Scienfitz replied:
regarding lines, wasnt able to add some nice lines this is best I could do 
<img width="1129" height="312" alt="image" src="https://github.com/user-attachments/assets/9d6a7c08-6bc4-4d5c-8c5d-8bf84e736f28" />
is that better in your view?

AVHopp replied:
Re Lines: Then I actually prefer it without the lines, if they are not fully connected it looks even weirder imo.

Re your example: This is NOT the same. What you claim to be "extremely similar" is something different, and also not well defined. And defining a function like `f(x,y) = {x+y, if x>0 // 1, else` is perfectly fine (however not using the same notation again in the line defining that function, but this part is not the point here). My main issue is that you use the same function again in your notation for a subspace of the full space which is not well defined. 

Let's take a specific example: `f(x,y) = x^2+2x + (x^2-1)*y`. Then, one could argue that "if `x=1` or `x=-1`, then the value of `y` does not matter", but you can still not simply write down `f(x)` since `f` is a function defined on a 2D space, and `f(x)` would be defined on a 1D space and you cannot use the same notation for the function since it is not the same! Also, using `f(x,y)` to define `f(x,y)` is formally also not correct, but this is a point where I am defintely on your side - it is clear what is meant. One would thus need another function, e.g. `g(x)` which is well-defined on a 1D subspace.

TL;DR: I agree that we do not need the full formality here, but we need something that is at least reasonably well defined, what the current version is not. Let me also summon @kalama-ai and @AdrianSosic here for their opinion.

Scienfitz replied:
May I ask where in this section, code or PR I am trying to "define a function"? This here is merely pointing out what `f` depends on, not aiming to define. The words "functional definition" in the header describe the symmetry, NOT any function.

AVHopp replied:
You implicitly define `f` by how you write it here, at least partly.

My whole point is that you say `f(x,y)=f(x) under given circumstances` which is NOT correct as `f` is defined on a `X x Y` space so you cannot evaluate it just on the `X` space by just ignoring the `Y` part.

Scienfitz replied:
<img width="695" height="31" alt="image" src="https://github.com/user-attachments/assets/11f15ac3-8729-41db-9e4e-e72af5be114d" />

? of course I can and you already point out how in your own words, since it doenst depend on `y` I can pick any (or even no) element from `Y`


More examples with which there is absolutely *nothing* wrong:
1) $f(x,y) = x^2$
2) $f(x,y) = 1$ (contrary to what you say this is perfectly fine and follows the very same logic as 1)
3) since 1 and 2 are fine its also fine without explicitl funcitonal form, ie $f(x,y) = f(x)$. You can find this in many tetxbooks that aim to say "usually depends on `x` and `y` but in this special case just depends on `x`" which is pretty much 100% what I want to say here

AVHopp replied:
To not further escalate here, let's either discuss via chat and also hear what the others have to say here.

Scienfitz replied:
- renamed the partial functions
- removed the horizontal line, markdown doesnt seem to uspport hlines that go through completely

<img width="970" height="245" alt="image" src="https://github.com/user-attachments/assets/d155ecd5-271e-4818-84fd-3277e4c8dc49" />

AVHopp replied:
Muchas gracias :)

AVHopp commented on examples/Symmetries/permutation.py:
```diff
@@ -0,0 +1,214 @@
+# # Optimizing a Permutation-Invariant Function
+
+# In this example, we explore BayBE's capabilities for handling optimization problems
+# with symmetry via automatic data augmentation and / or constraint.
```
Any reason for the spaces around "/"? I'd prefer it without them.

Scienfitz replied:
changed

AVHopp commented on examples/Symmetries/permutation.py:
```diff
@@ -0,0 +1,214 @@
+    return df_z
+
+
+# Grid and dataframe for plotting
```
I think we do not need this comment as well as the "Plot the contour..." one in the next block

Scienfitz replied:
removed

AVHopp commented on baybe/utils/augmentation.py:
```diff
@@ -94,6 +98,69 @@ def df_apply_permutation_augmentation(
+
+        # Create the augmented row by mirroring the point at the mirror point.
+        # x_mirrored = mirror_point + (mirror_point - x) = 2*mirror_point - x
+        if row[column] != mirror_point:
```
Not 100% but do we want an exact inequality here or an approximate one? Feel free to decide what you think makes more sense, just want to raise awareness

Scienfitz replied:
well that is related tot he topic #779 how floats in general should be compared

Unless thats decided I'm not a fan to discuss this now in every PR separately

AVHopp replied:
Fine for me

AVHopp commented on baybe/recommenders/pure/bayesian/base.py:
```diff
@@ -86,6 +86,22 @@ def _get_acquisition_function(self, objective: Objective) -> AcquisitionFunction
+                template = surrogates.template
+                if isinstance(template, Surrogate):
+                    return template
+        return None
```
If I remember our logic correctly, then we allow users to implement whatever surrogates they want as long as they implement the `SurrogateProtocol`. In particular, we do not enforce them to implement the `Surrogate` class or inherit from this. This means that this function returns `None`, hence no augmentation will be done. However, I do not think that the user is informed at any point that no augmentation has been done in that case. Also, I think the same happens for a `CompositeSurrogate` that does not have a `_ReplicationMapping` but actual different surrogates.

Three questions:
1. Do I understand this logic here correctly?
2. Do we already have some sort of safe guard/warning informing the user that no augmentation will be done in this case?
3. If not, can/do we want to add one?

Scienfitz replied:
hmm in the case you describe there is also no moment when the user ever assingned any `symmetries` because their custom surrogate doesnt have this attribute - so why would there be an expectation of applied symmetries that has to be warned about?

AVHopp replied:
True, I overlooked where the symmetries live exactly!

AVHopp reviewed (approved):
All in all happy, only some comments but I do not see any major thing that needs change. Thank you for this cool new feature!

Scienfitz committed:
Improve example

Scienfitz committed:
Rename partial functions

Scienfitz commented:
@AdrianSosic appreciate your review

AdrianSosic commented on baybe/utils/conversion.py:
```diff
@@ -42,6 +42,18 @@ def nonstring_to_tuple(x: Sequence[_T], self: type, field: Attribute) -> tuple[_
     return tuple(x)
 
 
+def normalize_convertible2str_sequence(
```
My comment from [here](https://github.com/emdgroup/baybe/pull/626#discussion_r2532997248) is actually still unresolved:
* The name is super-hard to read and breaks our conventions (took me ages to understand that the part with the "2" is to be read as one piece)
* The function doesn't care at all about whether the content is bools/str, so why put it in the name
* Should mention that this is attrs-format

So how about:
1) Using a proper generic as input/output type of the sequence/tuple
2) rename the thing to `to_sorted_tuple`
3) Optionally turn docstrings into something like `Attrs-converter transforming sequences into sorted tuples` or similar?

Scienfitz replied:
replaced the whole logic with composable converters using the old `nonstring_to_tuple` and the new `sort_tuple`
6b2a4c74cf510b072515d026a310218856825831

AdrianSosic replied:
Perfect. For my taste this is 1) much easier to read since 2) just built from composable blocks (hence modular) 👍🏼

AdrianSosic commented on baybe/parameters/categorical.py:
```diff
@@ -38,8 +32,10 @@ class CategoricalParameter(_DiscreteLabelLikeParameter):
         alias="values",
-        converter=Converter(_convert_values, takes_self=True, takes_field=True),  # type: ignore
-        validator=(
+        converter=Converter(  # type: ignore[misc,call-overload]  # mypy: Converter
```
what is the `mypy` part about? Haven't seen this before

Scienfitz replied:
i think these are over eagerly LLM lines, the type ignore lines are already sufficient

removed the `mypy` lines

Scienfitz replied:
I now remember what the `mypy` lines were supoosed to be for: Make a comment why this ignore was needed (not always clear from the `type: ignore[X]` line)

but as we can see that has failed spectacularly :D

AdrianSosic commented on baybe/parameters/categorical.py:
```diff
@@ -38,8 +32,10 @@ class CategoricalParameter(_DiscreteLabelLikeParameter):
+        converter=Converter(  # type: ignore[misc,call-overload]  # mypy: Converter
+            normalize_convertible2str_sequence, takes_self=True, takes_field=True
+        ),
+        validator=(  # type: ignore[arg-type]  # mypy: validator tuple
```
why suddenly ignores needed (and what is the mypy part about)? Due to newer mypy release?

Scienfitz replied:
i removed the mypy part

without ignore I get: `baybe/parameters/categorical.py:38: error: Argument "validator" to "field" has incompatible type "tuple[Callable[[Any, Any, list[Any]], None], Callable[[Any, Attribute[tuple[str | bool, ...]], tuple[str | bool, ...]], Any]]"; expected "Callable[[Any, Attribute[tuple[str | bool, ...]], tuple[str | bool, ...]], Any] | Sequence[Callable[[Any, Attribute[tuple[str | bool, ...]], tuple[str | bool, ...]], Any]] | None"  [arg-type]`

AdrianSosic replied:
good that we talked about it: 93c0f74cfb2f0a02539a4af8913edbb2f9b420d6

AdrianSosic commented on baybe/parameters/numerical.py:
```diff
@@ -13,9 +13,10 @@
+from baybe.parameters.validation import validate_unique_values
 from baybe.settings import active_settings
 from baybe.utils.interval import InfiniteIntervalError, Interval
+from baybe.utils.validation import validate_is_finite
```
Question: this PR moves some stuff around (like the `validate_is_finite` and others), which technically represents breaking changes. Changelog entry or no?

Scienfitz replied:
no
moving mere validators is not a metnionworthy breaking change...

AdrianSosic commented on baybe/symmetries/__init__.py:
```diff
@@ -0,0 +1,13 @@
+    "MirrorSymmetry",
+    "PermutationSymmetry",
+    "Symmetry",
+]
```
For other components, we don't expose the base class in the namespace. So for consistency, I suggest:
```suggestion
from baybe.symmetries.dependency import DependencySymmetry
from baybe.symmetries.mirror import MirrorSymmetry
from baybe.symmetries.permutation import PermutationSymmetry

__all__ = [
    "DependencySymmetry",
    "MirrorSymmetry",
    "PermutationSymmetry",
]
```

Scienfitz replied:
fixed

AdrianSosic commented on baybe/constraints/conditions.py:
```diff
@@ -95,7 +95,7 @@ class Condition(ABC, SerialMixin):
 
     Conditions always evaluate an expression regarding a single parameter.
-    Conditions are part of constraints, a constraint can have multiple conditions.
+    Conditions are part of constraints and symmetries.
```
Maybe links to `constraints` and `symmetries` would be good?

Scienfitz replied:
added

AdrianSosic commented on baybe/constraints/discrete.py:
```diff
@@ -271,39 +270,56 @@ def get_invalid(self, data: pd.DataFrame) -> pd.Index:
 
     @override
     def get_invalid(self, data: pd.DataFrame) -> pd.Index:
-        # Get indices of entries with duplicate label entries. These will also be
```
It's been a while .... Can you maybe quickly comment on the reason for the logic change again?

AdrianSosic replied:
I guess this is related to the "diagonal" points you mention in the Changelog, tho I do not understand what "diagonal" really means here

Scienfitz replied:
When I was implementing the permutaiton constraint I was only thinking of the mixtue use case. In that case you would obviously always remove duplciated labels as well. For some reaosn my brain connected these two concepts and so I made the linked parameter constraint part of the permutaiton constraint. But in reality it should be fully independent and not hardcoded included, it just belongs to the list of cosntraints needed for properly modeling a mixture

I realized this when I crafted the new example: 2 discrete parameters. Imagine the searchspace as a 2D matrix. When we make them permutaiton invariant we are removing the upper (or equivalently lower) triangle of that matrix. But NOT the diagonal, which has nothing to do with permutaiton invariance, (1,1) is a degernate pointbut it has no permutation clone so nothing on the diag should be removed. This showed up in the number of points removed not matchign the expecttation: diagonal was removed due to the linked parameter cosntraint, but shouldnt.

AdrianSosic replied:
Ok, got it. Then two follow-ups:
1) Are there any places (e.g. in docs, examples, whatsoever) where we now need to account for the semantics change, i.e. add back an additional duplicates constraint?
2) I understand your explanation of `diagonal` in your 2D example (lattice of points), but I feel that no one would understand the changelog entry without this context. Like me, people would wonder `diagonal with respect to what???` 😄 Maybe you have a slightly biased view here because you worked on it. Can you perhaps rephrase the entry?

Scienfitz replied:
1) as far as I can see no. Most importantly: examples already sued the linked parameter constraints separately (so secretly applied it twice) which however now means they do not need to be changed at all
2) changed the changelog

AdrianSosic commented on baybe/recommenders/pure/bayesian/base.py:
```diff
@@ -86,6 +86,22 @@ def _get_acquisition_function(self, objective: Objective) -> AcquisitionFunction
+        """Get the Surrogate instance for augmentation/validation, if available."""
+        from baybe.surrogates.composite import CompositeSurrogate, _ReplicationMapping
+
+        model = self._surrogate_model
```
I'm not sure I understand the downstream consequences of this branching logic 🤔 
1. What happens is we have a `CompositeSurrogate` that is *not* a `_ReplicationMapping`? The method would return `None` in this case!?
2. What's the reason for the `isinstance(template, Surrogate)` check?
3. I'm not yet convinced by the `None` output type. This function is only ever called in places where a surrogate is (should!) be be available. So I think replacing the `None` path with an `assert self._surrogate is not None` check is the better approach

Scienfitz replied:
the branch checks these two subclasses and its not there to check whether the surroage tis avaialble butw hether a surrogate that can augment is avaialble


the logic used here is that those 2 are the only paths the surrogate protocol that have the `augment_measurements` method guaranteed that can be used for augmentation, others generally dont have it (not enforced in protocol). 

Now how would you improve that? `hasattr` check? A new protocol `AugmentationSurrogate` derived from the base protocol?

AdrianSosic replied:
Ah, thanks for explaining. I now understand that the role of this logic is essentially to trace `all surrogates capable of augmentation` right? 

However, I think this uncovers a fundamental problem of the way responsibilities are currently assigned, and gives rise to a very brittle construction. Before I explain, I think you agree that this essentially is a maintenance nightmare because we need to *remember* to expand this logic whenever something changes in the surrogate class hierarchy, i.e. when we add (for whatever reason) another such class or similar? Also, and maybe more importantly from the algorithmic view, it means that augmentation would *never* happen in case of custom surrogates, but augmentation is really a surrogate-agnostic mechanism!

In essence, I think the problem arises because of a wrong direction of responsibility. So far, we have two ways symmetries are handled:
* Using data augmentation, which *should* be completely surrogate-agnostic
* In the yet-to-come model-based way.

If I'm not mistaken, the problem is that the current code achieves the former via the latter, i.e. we let the surrogate handle the data augmentation. This causes the spurious coupling of the objects, which then requires the `is None` check. The problem would not arise if we strictly decoupled these two things, which essentially are two orthogonal mechanisms.

Or am I overlooking something here?

Scienfitz replied:
I think you're spot on tbh, we can agree this is the worst logic bit remaining in the PR

remaining question is, if the augmentation is surrogate-independent (which I agree with):
- We can move the caller of the augmentation to a higher-level object (in this case the recommender would make sense). The code bit ehre that extracts the model would be completely obsolete as the surrogates no longer have the `augment_` method. So thats the easy bit
- But we still need a way to turn the augmentation on/off. If we move this flag then also away from the models I feel like this would be weird

The whole concept of this PR is that the two aspects you mentioned above are consolidated in one Symmetry concept. Do we really need to make it more complicated?

Scienfitz replied:
changed due to the new agreed architecture discussed, resolving here

AdrianSosic commented on baybe/surrogates/base.py:
```diff
@@ -90,6 +92,14 @@ class Surrogate(ABC, SurrogateProtocol, SerialMixin):
     compatible."""
 
+    symmetries: tuple[Symmetry, ...] = field(
+        factory=tuple,
```
```suggestion
        default=(),
```

Scienfitz replied:
1. This project uses `factory=tuple` in >1 places and `default=()` in 0 places
2. 
<img width="617" height="365" alt="image" src="https://github.com/user-attachments/assets/0e132879-dcd2-4ffa-b519-7932a038b1de" />

AdrianSosic commented on baybe/surrogates/base.py:
```diff
@@ -115,6 +125,27 @@ class Surrogate(ABC, SurrogateProtocol, SerialMixin):
+
+        Returns:
+            A dataframe with the augmented measurements, including the original
+            ones.
```
```suggestion
            A dataframe with the augmented measurements, including the original ones.
```

Scienfitz replied:
changed

AdrianSosic commented on baybe/symmetries/base.py:
```diff
@@ -0,0 +1,89 @@
+    def augment_measurements(
+        self,
+        measurements: pd.DataFrame,
+        parameters: Iterable[Parameter] | None = None,
```
Looking at your code, I think `Sequence` is probably the safer choice here, given that you already (unnoticedly) broke the `Iterable` contract by iterating multiple times over it 😬 What do you think?

```suggestion
        parameters: Sequence[Parameter] | None = None,
```

Scienfitz replied:
changed

AdrianSosic commented on baybe/symmetries/base.py:
```diff
@@ -0,0 +1,89 @@
+            measurements: The dataframe containing the measurements to be
+                augmented.
+            parameters: Optional parameter objects carrying additional information.
+                Only required by specific augmentation implementations.
```
```suggestion
            measurements: The dataframe containing the measurements to be augmented.
            parameters: Optional parameter objects carrying additional information.
                Only required by some symmetry classes.

Scienfitz replied:
changed

AdrianSosic commented on baybe/symmetries/base.py:
```diff
@@ -0,0 +1,89 @@
+                f"The symmetry of type '{self.__class__.__name__}' was set up with the "
+                f"following parameters that are not present in the search space: "
+                f"{parameters_missing}."
+            )
```
A bit shorter
```suggestion
        if missing := set(self.parameter_names) - set(searchspace.parameter_names):
            raise IncompatibleSearchSpaceError(
                f"The symmetry of type '{self.__class__.__name__}' was set up with the "
                f"following parameters that are not present in the search space: "
                f"{missing}."
            )
```

Scienfitz replied:
change

AdrianSosic commented on CHANGELOG.md:
```diff
@@ -15,16 +15,23 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 - Interpoint constraints for continuous search spaces
 - Transfer learning benchmarks for shifted and inverted Hartmann functions
 - Coding convention instructions for agentic developers (`AGENTS.md`, `CLAUDE.md`)
+- Symmetry classes (`PermutationSymmetry`, `MirrorSymmetry`, `DependencySymmetry`)
```
maybe add the base class? You could even mention the symmetry framework as a whole first and only then mention the classes, because that the former is new is not clear from the bullet (you cannot distinguish it from the case where we later add a fourth class)

Scienfitz replied:
I changed it but ngl it reads super duplicated...
65977af6009776f8d8b5d63bb4b1bf1647199108

AdrianSosic replied:
Yeah, I agree, I was thinking more in terms of a single bullet like
```
- `Symmetry` concept for expressing symmetries of the optimization problem, including
  `PermutationSymmetry`, `MirrorSymmetry`, `DependencySymmetry` classes

Scienfitz replied:
I changed the changelog

btw if you are such very particular with rephrasing then it is ideal and saves time if you already post your suggestion in the original comment

AdrianSosic reviewed (changes requested):
Gut Ding will Weile haben ...

AdrianSosic commented on baybe/symmetries/dependency.py:
```diff
@@ -0,0 +1,169 @@
+    condition: Condition = field(validator=instance_of(Condition))
+    """The condition specifying the active range of the causing parameter."""
+
+    affected_parameter_names: tuple[str, ...] = field(
```
I'm still puzzled why you chose the terminology `causing` and `affected` when the the class is named  `Dependent...` and you talk about `dependent parameters` in the docstring. Wouldn't it much more sense to call the attribute `dependent_...` and correspondingly use `independent` for the other parameter types, which furthermore avoids the clash with the actual `causal` terminology from `causal modeling`?

Scienfitz replied:
I'm not willing to discuss these terminologies as part of this PR
pleas redirect this discussion and/or your ideas to #670

AdrianSosic commented on baybe/symmetries/dependency.py:
```diff
@@ -0,0 +1,169 @@
+    """The condition specifying the active range of the causing parameter."""
+
+    affected_parameter_names: tuple[str, ...] = field(
+        converter=Converter(  # type: ignore[misc,call-overload]  # mypy: Converter
```
Again: what is this `mypy` comment?

Scienfitz replied:
yep, LLM residual, not needed and removed

AdrianSosic commented on baybe/symmetries/dependency.py:
```diff
@@ -0,0 +1,169 @@
+    """The parameters affected by the dependency."""
+
+    n_discretization_points: int = field(
+        default=3, validator=(instance_of(int), ge(2)), kw_only=True
```
I can unterstand that you want to make the concept useable in conti spaces, but I would suggest to *not* chose any default then. I think we all agree that there is no obvious "default" that one could possibly chose to well-approximate a continuous function with discrete points, and the switch from disc to conti is one that moves from an exact representation of the symmetry to an approximate one. So I think this is not something that should be hidden from the user but where they should explicitly opt-in, don't you agree?

Scienfitz replied:
but here IS an obvious default?

2 points (to cover the bounds obviously) or 3 to cover one variation in between. This is same logic as in DOE

AdrianSosic replied:
Yes, and we all argue that DOE is bullshit in our talks for exact this reason 😬  All I'm saying – and I guess @AVHopp  and @kalama-ai might agree here – is that I consider it bad practice to switch from exact to approximate behind the scenes because many user will not realize that they cross this boundary and that this a moment when a real performance-vs-complexity tradeoff needs to be made

Scienfitz replied:
I hope you dint think DOE is bad because they are doing the 2 point parameter range thing? In the premise of the theory its simply a reasonable thing to do...

But I can live with the error prompting the user to do additonal work instead
432975b792ae908a2d106d0e3391d7bbe81679cf

AdrianSosic commented on baybe/symmetries/dependency.py:
```diff
@@ -0,0 +1,169 @@
+        parameters: Iterable[Parameter] | None = None,
+    ) -> pd.DataFrame:
+        # See base class.
+        if not self.use_data_augmentation:
```
Since this check *has* to happen in each subclass, let's perhaps turn it into the template pattern?

Scienfitz replied:
done here 72016153032c3e8ec394228ac8b436cc72a64e2f

AdrianSosic replied:
Good 👍🏼  Maybe last suggestions: let's perhaps not copy-paste the docstrings (for maintainability) but simply point the the public class for certain sections? And if possible, let's avoid hardcoding the `use_data_augmentation` name?

Scienfitz replied:
ive used the same pattern as applied to `get_invalid` and `_get_invalid`: docstring ion both baseclass strings and an additional note in the `_*` one

Scienfitz replied:
I also removed the hardcoded name reference

AdrianSosic commented on baybe/symmetries/dependency.py:
```diff
@@ -0,0 +1,169 @@
+        # values that are not active, as rows containing them should be
+        # augmented.
+        param = next(
+            cast(DiscreteParameter, p)
```
Why a cast? I think this needs to be a proper validation instead, no?

Scienfitz replied:
so you mean with `isisntance` or `is_discrete` checkcs? Or could I simply make the type hint of the method more preciese (overwriting the base class but I guess thats not an issue)

Scienfitz replied:
im resolving this thread as essential duplicate of https://github.com/emdgroup/baybe/pull/626#discussion_r3451255430

AdrianSosic commented on baybe/symmetries/dependency.py:
```diff
@@ -0,0 +1,169 @@
+        affected: list[tuple[str, tuple[float, ...]]] = []
+        for pn in self.affected_parameter_names:
+            p = next(p for p in parameters if p.name == pn)
+            if p.is_discrete:
```
I think that – as long as our `is_discrete` method does not work properly as a type guard (hope `ty` enables this!) – building branching logic based on it us both ugly and brittle. 
* What do we save: nothing (i.e. writing `p.is_discrete` instead of `isinstance(p, DiscreteParameter`)`. 
* What do we lose: type safety, and having to add all sorts of casts and ignores.

I suggest: let's change it

Scienfitz replied:
no point to pile that on top of the heap of things this PR already changes (this pattern also appear in many places unrelated to the PR)

please change it in a more related PR like eg #845

AdrianSosic commented on baybe/symmetries/dependency.py:
```diff
@@ -0,0 +1,169 @@
+                f"The symmetry of type '{self.__class__.__name__}' was set up "
+                f"with at least one parameter which is not present in the "
+                f"search space: {parameters_missing}."
+            )
```
This is done in the base class
```suggestion
```

Scienfitz replied:
yes and no

yes because the base class valdiates based ont he peroperty `parameter_names`

no because this class (for some reason) did not include the affected parameetersin that property yet

I fixed the latter and removed the valdiation as suggested

AdrianSosic commented on baybe/symmetries/dependency.py:
```diff
@@ -0,0 +1,169 @@
+            )
+
+        # Causing parameter must be discrete
+        param = searchspace.get_parameters_by_name(self._parameter_name)[0]
```
IMO the validation flow is not yet idea. I understand that you want to have validation in one single place, but from the class' perspective, this validation step is never called, which is also the reason why you require the `cast` in `augment_measurements` (just a side note: in the current implementation, the `cast` should rather be an `assert`). But this also means that I get no validation if I simply call `augment_measurements` on a `DependencyConstraint` object that exists in isolation, which is not great. 

The solution is perhaps to simply call the validation as part of `augment_measurements`, and not in the surrogate base class!?

Scienfitz replied:
of course a bare  cosntraint does not have the searchspace conetxt, so how am I suppsoed to validate anything that relates to other obejcts in the searchspace?

The only possibility I see is to extrract a valdiation utility that is used in BOTH `augment_` (at runtime) and `valdiate_searshcpace_` (whenever higher objects using this constraint are composed), but then we again have duplicated validation...

I used `cast` as an alternative way to `assert` for mypy but I can change it to that of course

AdrianSosic replied:
First observation: even in the current case, you only get the validation at recommendation time (aka run time) because that's when the recommender first sees the search space. So we do not lose anything if we move the validation to be executed in the `Symmetry` class instead. 

Second observation: if I'm not mistaken, all your validations currently only use the passed search space to extract parameter from it. This raises the question if the validation function should not simply accept a collection of parameters instead. If you do that, you have all ingredients in place to simply run the validation as part of `_augment_measurements` itself, potentially resolving the `cast` situation. Alternative, if you have a good reason to keep the input elevated to `SearchSpace` (please share), then instead it probably makes sense to also elevate `_augment_measurements` itself, because they logically/semantically should operate on the same level – which again allows to move the validation there.

Thoughts?

Scienfitz replied:
Just so were on the same page: the `validate_X_context` method is there so it can be called by a higher-level object once both are available while having the validation logic still inside the low-level class itself (makes sense to me). This by definition only helps the stateful case because in the stateless case only at runtime of `.recommend` all required objects come together (<- I hope you're not criticizing this because its really by definition and not the fault of this PR whatsoever)

on 1) This is simply because the campaign currently does not use utility functions to validate objects in the manner proposed above. But its totally possible that it would call a `Recommender.valdiate_` utility which in turn calls the `Symmetry.validate_` utils
- Then you would have a proper ahead-valdiation and the funciton youre trying to get rid of has a clear purpose. - In addition `augment_` can simply also call `validate_` but it would mean double validaiton (not problematic imo)
- Since the basic recommender protocol does not have the validations we would need to derive a new one OR use `hasattr` for this in the Campaign code

2) open to this trivial change but the other parts have to be decided first

Scienfitz replied:
changed due to the new agreed architecture discussed, resolving here

AdrianSosic commented on baybe/symmetries/mirror.py:
```diff
@@ -0,0 +1,86 @@
+
+    # object variables
+    mirror_point: float = field(
+        default=0.0, converter=float, validator=validate_is_finite, kw_only=True
```
kw_only not necessary IMO
```suggestion
    mirror_point: float = field(
        default=0.0, converter=float, validator=validate_is_finite
```

Scienfitz replied:
removed

AdrianSosic commented on baybe/utils/augmentation.py:
Let's take this opportunity to make all the `df` function arguments in file positional-only

Scienfitz replied:
done

AdrianSosic commented on baybe/utils/augmentation.py:
```diff
@@ -8,24 +8,29 @@
 def df_apply_permutation_augmentation(
     df: pd.DataFrame,
-    column_groups: Sequence[Sequence[str]],
+    permutation_groups: Sequence[Sequence[str]],
```
I vaguely remember you and @AVHopp having a discussion about this but: do we really want to change the semantics of this function? I know there is a changelog entry for it, but still: this is strictly speaking the worst case of breaking change. The function has the same name, same number of arguments, and same argument types, same return type, but will silently do something very different from the original version --> silent bug. 

I'm asking because there is no real need for this change in the first place!?

Scienfitz replied:
so without this chagne we have a verys trange difference between how the epormtuations are passed to the symmetry and how they are passed ot the utility -> bad

in the discussion there was considerable touble to udnerstand the odlw ay how this function expectss arguemtns and what it does, the nwe version seems better andd I tend to agree

so overal I msut say tis a desriable chagne

I dont care too much about changing this function - it is not by any menas a very central and very frequently sued function (especially not explicitly sued), a breaking chagne entry can be done, would that not suffice?

AdrianSosic replied:
OK, if the two of you agree here, I'll just follow.

Perhaps to through just one idea into the room: the *silent* breaking change could be avoided if we used a new name pattern for the augmentation utilities altogether. Now that we have three of them, maybe there is even something more concise than the `df_apply_*_augmentation` name pattern? Maybe simply `augment_by_mirroring`, `augment_by_permutation`, and `augment_by_dependency`? Just some thoughts...

Scienfitz replied:
To be clear: this is a mere utility which (until this PR) was not even connected to any of the main components of the package
given this context and considering the age and size of this development, imo none of such changes are needed

AdrianSosic commented on baybe/utils/augmentation.py:
This is probably minor, but still: Do we want to make any claims about the ordering of dataframe content produced by the utilities in this function? Right now, some append the augmented values, while others squeeze them in.

Scienfitz replied:
can you rephrase? no idea what you mean? make claims where, in docstrings?

AdrianSosic replied:
I mean the order of the resulting dataframe rows. Do we promise a certain ordering (e.g. 1. *all* original points come first and then *all* augmentations or 2. augmentations of a point come immediately after that point) or do we simply make no guarantees on the resulting ordering? In any case, perhaps a short sentence in the "returns" section does not hurt?

Scienfitz replied:
no we don't promise any order and iirc the order depends on the method

since augmentation provides a set of (in a sense) equivalent points, why should be promise any specific order or even comment on it?

no specific care has been taken of the ordering in the implementations for the same reason

AdrianSosic commented on baybe/utils/dataframe.py:
```diff
@@ -192,6 +192,10 @@ def create_fake_input(
+
+    Note:
+        This function does not consider constraints and might provide unexpected or
+        invalid data if certain constraints are present.
```
IMO this is more confusion than not having it. On this level, the concept of `constraints` does not even exist. The situation would be different if one of the arguments was a `SearchSpace` and the function was silently ignoring its constraints, but this is not the case here
```suggestion
```

Scienfitz replied:
so let me get that right: You really think tis better to

- assume a user auto-infers that this method doesnt handle constraints (because it doesnt have any related arguments)

rather than
- making this explicit with this note that costs 3 lines of comments?

This utility willmsot likely be called in contexts where constriants might indeed be present (eg a campaign is created for a test and need fake measurements enow tc)

AdrianSosic replied:
To me it's rather that the note is providing details on the relationship to another concept that does not even exist in this context and is hence answering a question that hasn't even been asked. I also wouldn't mention any potential implications of using the produced fake data when fitting a surrogate on it, or similar.

IMO, the better version of a such a note would be one that explains it without mentioning `Constraints`, e.g. simply saying that the data is being created by treating the parameters in *independently*, hence not considering any potential implicit relationships.

But if you disagree, feel free to resolve

Scienfitz replied:
note removed

AdrianSosic commented on baybe/utils/validation.py:
```diff
@@ -261,3 +262,34 @@ def preprocess_dataframe(
+        )
+
+
+def validate_unique_values(  # noqa: DOC101, DOC103
```
We now have two `validate_unique_values` functions. I think you forgot to delete the existing one?

Scienfitz replied:
right I consolidated it, keeping the util in the general utils place 11ac80e6c91ee59e0f68ec606b229287087c24bd
but it required a lazy improt elsewhere to avoid circ import 
cdb67a488271efeef066bf3a535009e550b52c6e

AdrianSosic reviewed (changes requested):
And some more

Scienfitz committed:
Merge main

Scienfitz committed:
Fix merge issues in PermutationInvariance constraint

Prevent partial permutation dedup when dependencies are present, since
the dependency logic changes which rows are equivalent. Add a defensive
_can_evaluate guard to the dependency block in _get_invalid. Sort
constraints in the cartesian product test to match production ordering.

Scienfitz committed:
Remove redundant type ignores

Co-authored-by: AdrianSosic <adrian.sosic@merckgroup.com>

Scienfitz committed:
Remove top level import of abstract base class

Scienfitz committed:
Add docstring links

Scienfitz committed:
Use Sequence instead of Iterable for parameters in augment_measurements

Scienfitz committed:
Improve docstrings

Co-authored-by: AdrianSosic <adrian.sosic@merckgroup.com>

Scienfitz committed:
Remove kw_only

Co-authored-by: AdrianSosic <adrian.sosic@merckgroup.com>

Scienfitz committed:
Make df positional-only in augmentation utility functions

Scienfitz committed:
Shorten statement

Co-authored-by: AdrianSosic <adrian.sosic@merckgroup.com>

Scienfitz committed:
Include affected parameters in DependencySymmetry.parameter_names

AdrianSosic committed:
Fix unnecessary type ignores

Scienfitz committed:
Remove default for n_discretization_points and validate at searchspace level

Scienfitz committed:
Replace normalize_convertible2str_sequence with composable converters

Scienfitz committed:
Update CHANGELOG.md

Scienfitz committed:
Make normalize_input_dtypes a lazy import in utils/validation

This breaks a circular import chain that prevents other modules from
importing validation utilities: utils/validation -> utils/dataframe ->
parameters/base -> parameters/__init__ -> parameters/categorical ->
utils/validation. The function is only used inside a single function
body, so the lazy import has no performance impact.

Scienfitz committed:
Consolidate validate_unique_values into utils/validation

Scienfitz committed:
Fix mypy errors from converter and validation refactoring

Scienfitz committed:
Use template method pattern for augment_measurements

AdrianSosic commented on tests/hypothesis_strategies/conditions.py:
```diff
@@ -0,0 +1,24 @@
+
+def threshold_conditions():
+    """Generate :class:`baybe.constraints.conditions.ThresholdCondition`."""
+    return st.builds(ThresholdCondition, threshold=finite_floats())
```
What about the other attribute, i.e. operator and tolerance? (I know it's only been moved, but I'm wondering why we didn't add it)

Scienfitz replied:
added 5efdf9c7ab6e6b3635f85ca1765eab1213181dcd

AdrianSosic commented on docs/components/symmetries.md:
```diff
@@ -0,0 +1,51 @@
+# Symmetry
+{class}`~baybe.symmetries.base.Symmetry` is a concept tied to the structure of the searchspace.
+It is thus closely related to a {class}`~baybe.constraints.base.Constraint`, but has a
+different purpose in BayBE. If the searchspace is symmetric in any sense, you can
```
I think it would be really helpful to mention a brief example case and explain (e.g. with two bullets) how it would manifest in form of a constraint versus in form of a symmetry, and what the implications are.

Scienfitz replied:
the specific example is quite literally linked just below

If I read these exact lines here they already completely include what you request:
- if searchspace is symmetric ->
  - can exclude parts of it via constraints
  - to change the model and incorporate this knowledge the concept of Symmetry exists

These are the two major high level things that come to mind when treating a symmetric problem. It is not a matter of `versus`

AdrianSosic replied:
Yes, I'm not asking to rewrite or add much content. What I suggest is to simply make the information jump to the eye more prominently, which "fließtext" is usually not the best medium for. Since you are comparing one concept against the other, a structured text form can help: perhaps just two bullets where you mention in the first one what a constraint does (not) do and in the second how the symmetry comes to the rescue.

Something like:
---
... but has a different purpose in BayBE:
* **Constraint:** Excludes parts of the search space from recommendation. If a problem shows symmetries, you can use constraints to restrict the search to one of the redundant regions.
* **Symmetry:** ...

Scienfitz replied:
incorporated something

Scienfitz committed:
Remove hardcoded name reference

Scienfitz committed:
Remove note

Co-authored-by: AdrianSosic <adrian.sosic@merckgroup.com>

Scienfitz committed:
Expand threshold_conditions hypothesis strategy to vary all attributes

AdrianSosic commented on examples/Symmetries/permutation.py:
```diff
@@ -0,0 +1,212 @@
+
+# There are several ways to handle such symmetries. The simplest one is
+# to augment your data. In the case of permutation invariance, augmentation means for
+# each measurement $(x,y)$ you also add a measurement with switched values, i.e.,  $(y,x)§.
```
```suggestion
# each measurement $(x,y)$ you also add a measurement with switched values, i.e.,  $(y,x)$.
```

Scienfitz replied:
fixed

Scienfitz committed:
Remove use_data_augmentation flag from Symmetry

Scienfitz committed:
Change augment_measurements signature to take SearchSpace

Scienfitz committed:
Move data augmentation from surrogate to BayesianRecommender

Data augmentation is a recommender-level concern, not a surrogate concern.
The surrogate retains no augmentation logic. Symmetries on the surrogate
will be used for future architectural adjustments only.

Scienfitz committed:
Add private symmetries attribute to GaussianProcessSurrogate

Symmetries on the GP surrogate are reserved for future architecture
adjustments (e.g., invariant kernels). A NotImplementedError is raised
when symmetries are present until the architecture changes are in place.

Scienfitz committed:
Update tests for symmetry refactoring

Scienfitz committed:
Update examples and documentation for symmetry refactoring

Scienfitz committed:
Update CHANGELOG

Scienfitz committed:
Move n_discretization_points check to augmentation path

The check is augmentation-specific and should not be required when a
symmetry is only used for surrogate architecture adjustment.

Scienfitz committed:
Move data augmentation documentation to recommenders

Scienfitz committed:
Merge branch 'main' into feature/invariance_augmentation

# Conflicts:
#	CHANGELOG.md
#	tests/hypothesis_strategies/constraints.py

Scienfitz committed:
Fix math block

Scienfitz committed:
Adjust example

Scienfitz committed:
Improve text

Scienfitz committed:
Improve picture

Scienfitz commented:
@AdrianSosic changes we discussed are up, this PR should now be ready, if you want take a review of the cumulative last set of commits since our last discussion
Fork links in the PR (still building as I write this)

AdrianSosic reviewed (approved):
Alright, I think we're ready for merge now. For my taste, the added `_symmetry` attribute really brings no value, but if you prefer to keep it, leave it. Also, I'm still not 100% sure if the recommender is the better place to store the symmetries compared to the Search Space, but only time can show 🙃
